hoodie-client/src/main/java/com/uber/hoodie/io/compact/HoodieRealtimeTableCompactor.java

```diff
@@ -0,0 +1,169 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.uber.hoodie.io.compact;
+
+import com.uber.hoodie.WriteStatus;
+import com.uber.hoodie.common.model.HoodieAvroPayload;
+import com.uber.hoodie.common.model.HoodieRecord;
+import com.uber.hoodie.common.model.HoodieWriteStat;
+import com.uber.hoodie.common.table.HoodieTableMetaClient;
+import com.uber.hoodie.common.table.HoodieTimeline;
+import com.uber.hoodie.common.table.timeline.HoodieActiveTimeline;
+import com.uber.hoodie.common.table.timeline.HoodieInstant;
+import com.uber.hoodie.common.table.view.RealtimeTableView;
+import com.uber.hoodie.common.util.AvroUtils;
+import com.uber.hoodie.common.util.FSUtils;
+import com.uber.hoodie.common.util.HoodieAvroUtils;
+import com.uber.hoodie.config.HoodieWriteConfig;
+import com.uber.hoodie.exception.HoodieCommitException;
+import com.uber.hoodie.table.HoodieCopyOnWriteTable;
+import org.apache.avro.Schema;
+import org.apache.commons.collections.IteratorUtils;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.FlatMapFunction;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.api.java.function.PairFunction;
+import scala.Tuple2;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * HoodieRealtimeTableCompactor compacts a hoodie table with merge on read storage.
+ * Computes all possible compactions, passes it through a CompactionFilter and executes
+ * all the compactions and writes a new version of base files and make a normal commit
+ *
+ * @see HoodieCompactor
+ */
+public class HoodieRealtimeTableCompactor implements HoodieCompactor {
+    private static Logger log = LogManager.getLogger(HoodieRealtimeTableCompactor.class);
+
+    @Override
+    public HoodieCompactionMetadata compact(JavaSparkContext jsc, HoodieWriteConfig config,
+        HoodieTableMetaClient metaClient, RealtimeTableView fsView,
+        CompactionFilter compactionFilter) throws Exception {
+        // TODO - rollback any compactions in flight
+
+        String compactionCommit = startCompactionCommit(metaClient);
+        log.info("Compacting " + metaClient.getBasePath() + " with commit " + compactionCommit);
+        List<String> partitionPaths =
+            FSUtils.getAllPartitionPaths(metaClient.getFs(), metaClient.getBasePath());
+
+        log.info("Compaction looking for files to compact in " + partitionPaths + " partitions");
+        List<CompactionOperation> operations =
```

Comment:
Given this usage, CompactionTask is probably a better name than  CompactionOperation IMO